### PR TITLE
[pkg/stanza/adapter] Add ability to select one of multiple storage extensions

### LINKED ISSUE #10915: Support concurrent file storage extensions
**Is your feature request related to a problem? Please describe.**
The contrib has two storage extension implementations - [dbstorage](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage/dbstorage) and [filestorage](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage/filestorage), for which there are several use cases:
* [persistent buffering](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper#persistent-queue) in queued_retry (essentially any exporter)
* [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver), which when storage extensions is enabled will automatically use it for persisting status of file tailing

Currently, only one instance of the storage extension can be used, which is limiting the flexibility to have separate config for each of the purposes. The limitation is likely to avoid referencing the storage ID and instead relying on using the only available one ([example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/adapter/storage.go)) without the need to reference it, though it is also somewhat limiting

It was [brought during the review of enabling persistent queue capability](https://github.com/open-telemetry/opentelemetry-collector/pull/5457#discussion_r890268669) by default in the collector build. The proposal was to include additional property, where the specific instance could be references, e.g.

```yaml
  storage: file_storage/foo

...
extensions:
  file_storage/foo:
    directory: foo
```

**Describe the solution you'd like**

~Deprecate the current automated attachment of file storage extension to filelog receiver and `persistent_storage_enabled` flag of queued_retry.~ Provide new configuration property `storage` which could be used to specify which file storage extension should be used when more than one is available.

The (draft) [implementation for enabling persistent queue](https://github.com/open-telemetry/opentelemetry-collector/pull/5457) addresses this change for queued_retry

### PR DESCRIPTION
**Pull request by djaglowski**
Prior to this change, many log receivers could make use of storage extensions.
However, selecting a storage extension worked as follows:
1. If exactly one storage extension is configured, use it.
2. If no storage extensions are configured, do nothing.
3. If more than one storage extension is configured, fail.

The new behavior depends on a new 'storage' field, which the user can use to
indicate which storgae extension a receiver should use. This works as follows:
1. If the user does not specify the 'storage' field, behave the same as above.
2. If the user specifies 'storage: none', do nothing.
3. If the user specifies 'storage: foo', use 'foo'.
4. If 'foo' is not configured or is not a storage extension, fail.

Resolves #10915

### CODE REVIEW

**Review by atoulme — APPROVED** [review]
Maybe add a second storage extension in the readme?

**Comment by djaglowski** [comment]
There's been some additional discussion on the expected behavior [over here](https://github.com/open-telemetry/opentelemetry-collector/pull/5784#discussion_r936058961). New components are not going to use storage extensions by default. Due to this, we need to accept the breaking change (all affected components are still in alpha). I'm moving this PR back to draft status and will make the appropriate updates.

File: pkg/stanza/adapter/storage.go [hunk 1]
```diff
@@ -25,37 +26,64 @@ import (
...
-				}
-				storageExtension = se
-			}
+func GetStorageExtension(ctx context.Context, host component.Host, storageID config.ComponentID) (storage.Extension, error) {
+	if host == nil {
+		return nil, nil
+	}
+
+	// Storage explicitly disabled. (e.g. 'storage: false')
+	if storageID.String() == "false" {
```
Comment (tigrannajaryan) [inline pkg/stanza/adapter/storage.go:35]: We probably need to add somewhere in the core a check that "false" is not a valid extension type.

Also, should this be "false" or "none" as the PR description suggests?

**Comment by tigrannajaryan** [comment]
> This works as follows:
> 
> 1. If the user does not specify the 'storage' field, behave the same as above.
> 2. If the user specifies 'storage: none', do nothing.
> 3. If the user specifies 'storage: foo', use 'foo'.
> 4. If 'foo' is not configured or is not a storage extension, fail.

A further evolution of this could be:
5. If the user specified 'storage: *' then use the first available storage extension.

**Comment by tigrannajaryan** [comment]
I agree that explicitly specifying the storage is the right approach.

>How committed are you to the opinion you expressed about preserving the functionality, either for usability or backwards compatibility?

Are you referring to my proposal to allow `*` as the wildcard for storage name? I think this can be useful, but I am fine if we don't want to do it. I wasn't proposing it for backwards compatibility. I think it is OK to break it for an Alpha component. 

@dmitryax do you see any problem with this? I think this is used in our Helm Chart, so the chart may need to be updated.

**Comment by djaglowski** [comment]
Thanks @tigrannajaryan. I was referring to your comments on the related issue [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/10915#issuecomment-1154055913). Sounds like we are on the same page though. 👍

**Comment by djaglowski** [comment]
Closing in favor of #13418

**Pull request closed without merging**